File: extract_data.js
#!/usr/bin/env node
/**
 * Script para extrair dados do arquivo CSV dados_processados.csv
 * e gerar os arquivos JSON na mesma estrutura dos existentes.
 * 
 * Uso: node extract_data.js
 */

const fs = require('fs');
const path = require('path');

/**
 * Lê e processa o arquivo CSV
 */
function readCSVData(csvFilePath) {
    try {
        const fileContent = fs.readFileSync(csvFilePath, 'utf-8');
        const lines = fileContent.split('\n');
        
        if (lines.length < 2) {
            console.log('Arquivo CSV vazio ou sem dados');
            return [];
        }
        
        // Primeira linha são os headers
        const headers = lines[0].split(',').map(h => h.trim());
        console.log('Headers encontrados:', headers);
        
        const data = [];
        let processedCount = 0;
        let skippedCount = 0;
        
        for (let i = 1; i < lines.length; i++) {
            const line = lines[i].trim();
            if (!line) continue; // Pular linhas vazias
            
            try {
                // Dividir a linha respeitando aspas
                const values = parseCSVLine(line);
                
                if (values.length !== headers.length) {
                    console.log(`Linha ${i + 1}: Número incorreto de colunas (${values.length} vs ${headers.length})`);
                    skippedCount++;
                    continue;
                }
                
                // Criar objeto com base nos headers
                const row = {};
                headers.forEach((header, index) => {
                    row[header] = values[index] ? values[index].trim() : '';
                });
                
                // Validar campos obrigatórios
                if (!row['Local'] || !row['Local/End']) {
                    console.log(`Linha ${i + 1}: Local ou endereço vazio, pulando...`);
                    skippedCount++;
                    continue;
                }
                
                if (!row['Latitude'] || !row['Longitude']) {
                    console.log(`Linha ${i + 1}: Coordenadas vazias, pulando...`);
                    skippedCount++;
                    continue;
                }
                
                // Converter coordenadas
                const latitude = parseFloat(row['Latitude']);
                const longitude = parseFloat(row['Longitude']);
                
                if (isNaN(latitude) || isNaN(longitude)) {
                    console.log(`Linha ${i + 1}: Coordenadas inválidas, pulando...`);
                    skippedCount++;
                    continue;
                }
                
                // Converter número de estudantes
                const estudante = parseInt(row['Estudante']) || 0;
                
                // Montar registro no formato do locations.json
                const locationData = {
                    id: processedCount + 1,
                    local: row['Local'],
                    endereco: row['Local/End'],
                    disciplina: row['Disciplina'] || '',
                    preceptor: row['Preceptor'] || '',
                    estudante: estudante,
                    periodo_original: row['Período_Original'] || '',
                    turma: row['Turma'] || '',
                    categoria: row['Categoria'] || '',
                    turno: row['Turno'] || '',
                    latitude: latitude,
                    longitude: longitude
                };
                
                data.push(locationData);
                processedCount++;
                
            } catch (error) {
                console.log(`Erro na linha ${i + 1}: ${error.message}`);
                skippedCount++;
                continue;
            }
        }
        
        console.log(`\nProcessamento concluído:`);
        console.log(`  - Registros processados: ${processedCount}`);
        console.log(`  - Registros pulados: ${skippedCount}`);
        
        return data;
        
    } catch (error) {
        console.error(`Erro ao ler arquivo CSV: ${error.message}`);
        return [];
    }
}

/**
 * Parser simples para CSV que respeita aspas
 */
function parseCSVLine(line) {
    const values = [];
    let current = '';
    let inQuotes = false;
    
    for (let i = 0; i < line.length; i++) {
        const char = line[i];
        
        if (char === '"') {
            inQuotes = !inQuotes;
        } else if (char === ',' && !inQuotes) {
            values.push(current);
            current = '';
        } else {
            current += char;
        }
    }
    
    values.push(current); // Último valor
    return values;
}

/**
 * Extrai dados únicos para os filtros
 */
function extractFiltersData(locationsData) {
    const disciplinas = new Set();
    const periodos = new Set();
    const preceptores = new Set();
    
    locationsData.forEach(location => {
        if (location.disciplina) {
            disciplinas.add(location.disciplina);
        }
        
        if (location.periodo_original) {
            periodos.add(location.periodo_original);
        }
        
        if (location.preceptor) {
            preceptores.add(location.preceptor);
        }
    });
    
    const filtersData = {
        disciplinas: Array.from(disciplinas).sort(),
        periodos: Array.from(periodos).sort(),
        preceptores: Array.from(preceptores).sort()
    };
    
    console.log(`\nFiltros extraídos:`);
    console.log(`  - ${filtersData.disciplinas.length} disciplinas`);
    console.log(`  - ${filtersData.periodos.length} períodos`);
    console.log(`  - ${filtersData.preceptores.length} preceptores`);
    
    return filtersData;
}

/**
 * Salva dados em arquivo JSON
 */
function saveJSONFile(data, filePath) {
    try {
        fs.writeFileSync(filePath, JSON.stringify(data, null, 2), 'utf-8');
        console.log(`Arquivo ${filePath} salvo com sucesso!`);
        return true;
    } catch (error) {
        console.error(`Erro ao salvar ${filePath}: ${error.message}`);
        return false;
    }
}

/**
 * Compara novos dados com arquivo existente
 */
function compareWithExisting(newData, existingFile) {
    if (!fs.existsSync(existingFile)) {
        console.log(`Arquivo ${existingFile} não existe para comparação.`);
        return;
    }
    
    try {
        const existingData = JSON.parse(fs.readFileSync(existingFile, 'utf-8'));
        
        console.log(`\nComparação com ${existingFile}:`);
        console.log(`  - Registros existentes: ${existingData.length}`);
        console.log(`  - Novos registros: ${newData.length}`);
        console.log(`  - Diferença: ${newData.length - existingData.length}`);
        
        if (existingData.length > 0 && newData.length > 0) {
            console.log(`\nPrimeiro registro existente vs novo:`);
            console.log(`  Local: '${existingData[0].local}' vs '${newData[0].local}'`);
            console.log(`  Disciplina: '${existingData[0].disciplina}' vs '${newData[0].disciplina}'`);
        }
        
    } catch (error) {
        console.error(`Erro ao comparar com arquivo existente: ${error.message}`);
    }
}

/**
 * Função principal
 */
function main() {
    console.log('=== Script de Extração de Dados dos Estágios FAMINAS ===\n');
    
    // Caminhos dos arquivos
    const csvFile = 'public/dados_processados.csv';
    const locationsOutput = 'public/locations_new.json';
    const filtersOutput = 'public/filters_new.json';
    
    // Verificar se o arquivo CSV existe
    if (!fs.existsSync(csvFile)) {
        console.error(`Erro: Arquivo ${csvFile} não encontrado!`);
        return;
    }
    
    // Ler dados do CSV
    console.log('1. Lendo dados do arquivo CSV...');
    const locationsData = readCSVData(csvFile);
    
    if (locationsData.length === 0) {
        console.log('Nenhum dado válido encontrado no CSV!');
        return;
    }
    
    // Extrair dados de filtros
    console.log('\n2. Extraindo dados para filtros...');
    const filtersData = extractFiltersData(locationsData);
    
    // Salvar arquivos JSON
    console.log('\n3. Salvando arquivos JSON...');
    
    // Salvar locations
    if (saveJSONFile(locationsData, locationsOutput)) {
        compareWithExisting(locationsData, 'public/locations.json');
    }
    
    // Salvar filters
    if (saveJSONFile(filtersData, filtersOutput)) {
        compareWithExisting(filtersData, 'public/filters.json');
    }
    
    console.log(`\n=== Processamento concluído ===`);
    console.log(`Novos arquivos gerados:`);
    console.log(`  - ${locationsOutput}`);
    console.log(`  - ${filtersOutput}`);
    console.log(`\nPara substituir os arquivos atuais, execute:`);
    console.log(`  mv ${locationsOutput} public/locations.json`);
    console.log(`  mv ${filtersOutput} public/filters.json`);
}

// Executar se for chamado diretamente
if (require.main === module) {
    main();
}

module.exports = {
    readCSVData,
    extractFiltersData,
    saveJSONFile,
    compareWithExisting
};

File: extract_data.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Script para extrair dados do arquivo CSV dados_processados.csv
e gerar os arquivos JSON na mesma estrutura dos existentes.

Estrutura do CSV:
Semestre,Disciplina,Turma,Período,Grupos,Preceptor,Estudante,Categoria,Dia,Turno,Rodízio,Local,Local/End,Latitude,Longitude,Período_Original

Estrutura do locations.json:
{
  "id": int,
  "local": string,
  "endereco": string,
  "disciplina": string,
  "preceptor": string,
  "estudante": int,
  "periodo_original": string,
  "turma": string,
  "categoria": string,
  "turno": string,
  "latitude": float,
  "longitude": float
}

Estrutura do filters.json:
{
  "disciplinas": [string],
  "periodos": [string],
  "preceptores": [string]
}
"""

import csv
import json
import os
from typing import List, Dict, Set, Any

def validate_coordinates(latitude: float, longitude: float) -> bool:
    """
    Valida se as coordenadas estão dentro dos limites esperados para o Brasil.
    Brasil: Latitude entre -33.75 e 5.27, Longitude entre -73.98 e -28.84
    """
    # Limites aproximados do Brasil com margem de segurança
    if not (-35 <= latitude <= 6):
        return False
    if not (-75 <= longitude <= -25):
        return False
    return True

def format_coordinates(lat_raw: float, lon_raw: float) -> tuple:
    """
    Formata coordenadas do CSV para o formato correto.
    Ex: -1959565 -> -19.59565, -198846 -> -19.8846
    """
    # Se o valor já está no formato correto (entre -100 e 100), não modificar
    if -100 <= lat_raw <= 100 and -100 <= lon_raw <= 100:
        return lat_raw, lon_raw
    
    # Função para formatar uma coordenada individual
    def format_single_coord(coord):
        # Converter para string para analisar os dígitos
        coord_str = str(int(abs(coord)))
        
        # Determinar quantos dígitos devem ser decimais
        if len(coord_str) >= 7:  # -1959565 (7 dígitos) -> -19.59565
            decimal_digits = 5
        elif len(coord_str) == 6:  # -198846 (6 dígitos) -> -19.8846  
            decimal_digits = 4
        elif len(coord_str) >= 4:  # Outros casos
            decimal_digits = len(coord_str) - 2
        else:
            return coord  # Muito pequeno, não modificar
        
        divisor = 10 ** decimal_digits
        return coord / divisor
    
    latitude = format_single_coord(lat_raw)
    longitude = format_single_coord(lon_raw)
    
    return latitude, longitude

def read_csv_data(csv_file_path: str) -> List[Dict[str, Any]]:
    """
    Lê o arquivo CSV e retorna uma lista de dicionários com os dados.
    """
    data = []
    
    try:
        with open(csv_file_path, 'r', encoding='utf-8') as file:
            # Usar DictReader para mapear automaticamente as colunas
            reader = csv.DictReader(file)
            
            for row_num, row in enumerate(reader, start=2):  # Começar do 2 porque linha 1 é header
                try:
                    # Validar campos obrigatórios
                    if not row.get('Local') or not row.get('Local/End'):
                        print(f"Linha {row_num}: Local ou endereço vazio, pulando...")
                        continue
                    
                    if not row.get('Latitude') or not row.get('Longitude'):
                        print(f"Linha {row_num}: Coordenadas vazias, pulando...")
                        continue
                    
                    # Converter coordenadas para float com formatação correta
                    try:
                        # As coordenadas no CSV estão sem ponto decimal
                        # Ex: -1959565 deve ser -19.59565
                        lat_raw = float(row['Latitude'])
                        lon_raw = float(row['Longitude'])
                        
                        # Aplicar formatação correta das coordenadas
                        latitude, longitude = format_coordinates(lat_raw, lon_raw)
                        
                        # Validar se as coordenadas estão dentro dos limites do Brasil
                        # Ser mais flexível com a validação
                        if not validate_coordinates(latitude, longitude):
                            print(f"Linha {row_num}: Coordenadas possivelmente inválidas ({latitude}, {longitude}), mas incluindo...")
                            # Não pular, apenas avisar
                        
                    except (ValueError, TypeError):
                        print(f"Linha {row_num}: Erro ao converter coordenadas, pulando...")
                        continue
                    
                    # Converter número de estudantes para int
                    try:
                        estudante = int(row.get('Estudante', 0))
                    except (ValueError, TypeError):
                        print(f"Linha {row_num}: Erro ao converter número de estudantes, usando 0")
                        estudante = 0
                    
                    # Montar registro no formato do locations.json
                    location_data = {
                        "id": row_num - 1,  # ID sequencial começando do 1
                        "local": row['Local'].strip(),
                        "endereco": row['Local/End'].strip(),
                        "disciplina": row.get('Disciplina', '').strip(),
                        "preceptor": row.get('Preceptor', '').strip(),
                        "estudante": estudante,
                        "periodo_original": row.get('Período_Original', '').strip(),
                        "turma": row.get('Turma', '').strip(),
                        "categoria": row.get('Categoria', '').strip(),
                        "turno": row.get('Turno', '').strip(),
                        "latitude": latitude,
                        "longitude": longitude
                    }
                    
                    data.append(location_data)
                    
                except Exception as e:
                    print(f"Erro na linha {row_num}: {e}")
                    continue
    
    except FileNotFoundError:
        print(f"Arquivo {csv_file_path} não encontrado!")
        return []
    except Exception as e:
        print(f"Erro ao ler arquivo CSV: {e}")
        return []
    
    print(f"Total de registros processados: {len(data)}")
    return data

def extract_filters_data(locations_data: List[Dict[str, Any]]) -> Dict[str, List[str]]:
    """
    Extrai os valores únicos para os filtros a partir dos dados de localização.
    """
    disciplinas: Set[str] = set()
    periodos: Set[str] = set()
    preceptores: Set[str] = set()
    
    for location in locations_data:
        # Disciplinas
        if location.get('disciplina'):
            disciplinas.add(location['disciplina'])
        
        # Períodos
        if location.get('periodo_original'):
            periodos.add(location['periodo_original'])
        
        # Preceptores
        if location.get('preceptor'):
            preceptores.add(location['preceptor'])
    
    # Converter sets para listas ordenadas
    filters_data = {
        "disciplinas": sorted(list(disciplinas)),
        "periodos": sorted(list(periodos)),
        "preceptores": sorted(list(preceptores))
    }
    
    print(f"Filtros extraídos:")
    print(f"  - {len(filters_data['disciplinas'])} disciplinas")
    print(f"  - {len(filters_data['periodos'])} períodos")
    print(f"  - {len(filters_data['preceptores'])} preceptores")
    
    return filters_data

def save_json_file(data: Any, file_path: str, indent: int = 2) -> bool:
    """
    Salva dados em um arquivo JSON.
    """
    try:
        with open(file_path, 'w', encoding='utf-8') as file:
            json.dump(data, file, ensure_ascii=False, indent=indent)
        print(f"Arquivo {file_path} salvo com sucesso!")
        return True
    except Exception as e:
        print(f"Erro ao salvar {file_path}: {e}")
        return False

def compare_with_existing(new_data: List[Dict], existing_file: str) -> None:
    """
    Compara os novos dados com o arquivo existente para identificar diferenças.
    """
    if not os.path.exists(existing_file):
        print(f"Arquivo {existing_file} não existe para comparação.")
        return
    
    try:
        with open(existing_file, 'r', encoding='utf-8') as file:
            existing_data = json.load(file)
        
        print(f"\nComparação com {existing_file}:")
        print(f"  - Registros existentes: {len(existing_data)}")
        print(f"  - Novos registros: {len(new_data)}")
        print(f"  - Diferença: {len(new_data) - len(existing_data)}")
        
        # Verificar alguns campos específicos para identificar discrepâncias
        if len(existing_data) > 0 and len(new_data) > 0:
            print(f"\nPrimeiro registro existente vs novo:")
            print(f"  Local: '{existing_data[0].get('local')}' vs '{new_data[0].get('local')}'")
            print(f"  Disciplina: '{existing_data[0].get('disciplina')}' vs '{new_data[0].get('disciplina')}'")
            
    except Exception as e:
        print(f"Erro ao comparar com arquivo existente: {e}")

def test_coordinate_conversion():
    """
    Testa a conversão de coordenadas com exemplos conhecidos.
    """
    print("=== Teste de Conversão de Coordenadas ===")
    
    test_cases = [
        (-1959565, -4390824, "CAS Norte - Policlínica Regional Norte (7 dígitos)"),
        (-1986977, -4389453, "UBS Novo Alvorada (7 dígitos)"),
        (-198846, -437984, "UBS Siderúrgica (6 dígitos)"),
        (-1988232, -4389812, "UBS Nova Vista (7 dígitos)"),
        (-1978835, -439401, "UBS Catumbi (6 dígitos)"),
        (-1975626, -4384725, "UBS Bom Jesus (7 dígitos)")
    ]
    
    for lat_raw, lon_raw, local in test_cases:
        lat_formatted, lon_formatted = format_coordinates(lat_raw, lon_raw)
        is_valid = validate_coordinates(lat_formatted, lon_formatted)
        
        print(f"Local: {local}")
        print(f"  Original: ({lat_raw}, {lon_raw})")
        print(f"  Formatado: ({lat_formatted}, {lon_formatted})")
        print(f"  Válido: {'✓' if is_valid else '✗'}")
        print()

def main():
    """
    Função principal do script.
    """
    print("=== Script de Extração de Dados dos Estágios FAMINAS ===\n")
    
    # Executar teste de coordenadas
    test_coordinate_conversion()
    
    # Caminhos dos arquivos
    csv_file = "public/dados_processados.csv"
    locations_output = "public/locations_new.json"
    filters_output = "public/filters_new.json"
    
    # Verificar se o arquivo CSV existe
    if not os.path.exists(csv_file):
        print(f"Erro: Arquivo {csv_file} não encontrado!")
        return
    
    # Ler dados do CSV
    print("1. Lendo dados do arquivo CSV...")
    locations_data = read_csv_data(csv_file)
    
    if not locations_data:
        print("Nenhum dado válido encontrado no CSV!")
        return
    
    # Extrair dados de filtros
    print("\n2. Extraindo dados para filtros...")
    filters_data = extract_filters_data(locations_data)
    
    # Salvar arquivos JSON
    print("\n3. Salvando arquivos JSON...")
    
    # Salvar locations
    if save_json_file(locations_data, locations_output):
        # Comparar com arquivo existente
        compare_with_existing(locations_data, "public/locations.json")
    
    # Salvar filters
    if save_json_file(filters_data, filters_output):
        # Comparar com arquivo existente
        compare_with_existing(filters_data, "public/filters.json")
    
    print(f"\n=== Processamento concluído ===")
    print(f"Novos arquivos gerados:")
    print(f"  - {locations_output}")
    print(f"  - {filters_output}")
    print(f"\nPara substituir os arquivos atuais, renomeie:")
    print(f"  mv {locations_output} public/locations.json")
    print(f"  mv {filters_output} public/filters.json")

if __name__ == "__main__":
    main()
